diff -r -c GBrowse-2.48/htdocs/example_scripts/gbrowse_get_sessionid.pl GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_get_sessionid.pl
*** GBrowse-2.48/htdocs/example_scripts/gbrowse_get_sessionid.pl	2012-02-24 16:04:33.000000000 -0500
--- GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_get_sessionid.pl	2012-03-04 19:14:47.000000000 -0500
***************
*** 13,18 ****
  die $response->status_line unless $response->is_success;
  
  my $cookie = $response->header('Set-cookie');
! my ($id)   = $cookie =~ /gbrowse_sess=([a-f0-9]+)/;
  print $id,"\n";
  
--- 13,18 ----
  die $response->status_line unless $response->is_success;
  
  my $cookie = $response->header('Set-cookie');
! my ($id)   = $cookie =~ /gbrowse2_sess=([a-f0-9]+)/;
  print $id,"\n";
  
diff -r -c GBrowse-2.48/htdocs/example_scripts/gbrowse_new_sessionid.pl GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_new_sessionid.pl
*** GBrowse-2.48/htdocs/example_scripts/gbrowse_new_sessionid.pl	2012-02-24 16:04:33.000000000 -0500
--- GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_new_sessionid.pl	2012-03-04 19:15:18.000000000 -0500
***************
*** 13,18 ****
  die $response->status_line unless $response->is_success;
  
  my $cookie = $response->header('Set-cookie');
! my ($id)   = $cookie =~ /gbrowse_sess=([a-f0-9]+)/;
  print $id,"\n";
  
--- 13,18 ----
  die $response->status_line unless $response->is_success;
  
  my $cookie = $response->header('Set-cookie');
! my ($id)   = $cookie =~ /gbrowse2_sess=([a-f0-9]+)/;
  print $id,"\n";
  
diff -r -c GBrowse-2.48/htdocs/example_scripts/gbrowse_upload_data.pl GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_upload_data.pl
*** GBrowse-2.48/htdocs/example_scripts/gbrowse_upload_data.pl	2012-02-24 16:04:33.000000000 -0500
--- GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_upload_data.pl	2012-03-04 19:15:53.000000000 -0500
***************
*** 41,47 ****
  die $response->status_line unless $response->is_success;
  my $content     = $response->decoded_content;
  my $cookie      = $response->header('Set-cookie');
! my ($used_id)   = $cookie =~ /gbrowse_sess=([a-f0-9]+)/;
  
  my $struct  = from_json($content);
  
--- 41,47 ----
  die $response->status_line unless $response->is_success;
  my $content     = $response->decoded_content;
  my $cookie      = $response->header('Set-cookie');
! my ($used_id)   = $cookie =~ /gbrowse2_sess=([a-f0-9]+)/;
  
  my $struct  = from_json($content);
  
diff -r -c GBrowse-2.48/htdocs/example_scripts/gbrowse_upload_file.pl GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_upload_file.pl
*** GBrowse-2.48/htdocs/example_scripts/gbrowse_upload_file.pl	2012-02-24 16:04:37.000000000 -0500
--- GBrowse-2.48.eupath/htdocs/example_scripts/gbrowse_upload_file.pl	2012-03-04 19:13:40.000000000 -0500
***************
*** 35,41 ****
  die $response->status_line unless $response->is_success;
  my $content     = $response->decoded_content;
  my $cookie      = $response->header('Set-cookie');
! my ($used_id)   = $cookie =~ /gbrowse_sess=([a-f0-9]+)/;
  
  my $struct  = from_json($content);
  
--- 35,41 ----
  die $response->status_line unless $response->is_success;
  my $content     = $response->decoded_content;
  my $cookie      = $response->header('Set-cookie');
! my ($used_id)   = $cookie =~ /gbrowse2_sess=([a-f0-9]+)/;
  
  my $struct  = from_json($content);
  
diff -r -c GBrowse-2.48/htdocs/js/login.js GBrowse-2.48.eupath/htdocs/js/login.js
*** GBrowse-2.48/htdocs/js/login.js	2012-02-24 16:04:36.000000000 -0500
--- GBrowse-2.48.eupath/htdocs/js/login.js	2012-03-04 19:17:25.000000000 -0500
***************
*** 1207,1213 ****
      var currentUrl  = String(String(document.location).split('#')[0]).split('?')[0];
          currentUrl  = String(currentUrl).split('http://')[1];
          currentUrl  = currentUrl.slice(currentUrl.indexOf('/'));
!     document.cookie = 'gbrowse_sess=; max-age=0; path='+currentUrl;
  }
  
  //Retrieve the GET variables and pass them to the OpenID handler
--- 1207,1213 ----
      var currentUrl  = String(String(document.location).split('#')[0]).split('?')[0];
          currentUrl  = String(currentUrl).split('http://')[1];
          currentUrl  = currentUrl.slice(currentUrl.indexOf('/'));
!     document.cookie = 'gbrowse2_sess=; max-age=0; path='+currentUrl;
  }
  
  //Retrieve the GET variables and pass them to the OpenID handler
diff -r -c GBrowse-2.48/lib/Bio/Graphics/Browser2/Render.pm GBrowse-2.48.eupath/lib/Bio/Graphics/Browser2/Render.pm
*** GBrowse-2.48/lib/Bio/Graphics/Browser2/Render.pm	2012-02-24 16:04:36.000000000 -0500
--- GBrowse-2.48.eupath/lib/Bio/Graphics/Browser2/Render.pm	2012-03-04 19:16:23.000000000 -0500
***************
*** 64,70 ****
      ($data_source, $session) = @_;
    } elsif (@_ == 1) {
      my $globals = shift;
!     $requested_id = param('id')        || CGI::cookie('gbrowse_sess');
      $authority    = param('authority') || CGI::cookie('authority');
      my $shared_ok = Bio::Graphics::Browser2::Action->shared_lock_ok(param('action'));
      $session      = $globals->authorized_session($requested_id, 
--- 64,70 ----
      ($data_source, $session) = @_;
    } elsif (@_ == 1) {
      my $globals = shift;
!     $requested_id = param('id')        || CGI::cookie('gbrowse2_sess');
      $authority    = param('authority') || CGI::cookie('authority');
      my $shared_ok = Bio::Graphics::Browser2::Action->shared_lock_ok(param('action'));
      $session      = $globals->authorized_session($requested_id, 
diff -r -c GBrowse-2.48/lib/Bio/Graphics/Browser2/Session.pm GBrowse-2.48.eupath/lib/Bio/Graphics/Browser2/Session.pm
*** GBrowse-2.48/lib/Bio/Graphics/Browser2/Session.pm	2012-02-24 16:04:33.000000000 -0500
--- GBrowse-2.48.eupath/lib/Bio/Graphics/Browser2/Session.pm	2012-03-04 19:16:58.000000000 -0500
***************
*** 35,41 ****
  	  = @args{'driver','id','args','source','lockdir','locktype','expires','mode'};
  
  	$mode ||= 'exclusive';
! 	$CGI::Session::NAME = 'gbrowse_sess';     # custom cookie
  
  	#Probably safe to uncomment, but must test exhaustively with mysql locking
  	#to determine.
--- 35,41 ----
  	  = @args{'driver','id','args','source','lockdir','locktype','expires','mode'};
  
  	$mode ||= 'exclusive';
! 	$CGI::Session::NAME = 'gbrowse2_sess';     # custom cookie
  
  	#Probably safe to uncomment, but must test exhaustively with mysql locking
  	#to determine.
diff -r -c GBrowse-2.48/lib/Legacy/Graphics/Browser/PageSettings.pm GBrowse-2.48.eupath/lib/Legacy/Graphics/Browser/PageSettings.pm
*** GBrowse-2.48/lib/Legacy/Graphics/Browser/PageSettings.pm	2012-02-24 16:04:37.000000000 -0500
--- GBrowse-2.48.eupath/lib/Legacy/Graphics/Browser/PageSettings.pm	2012-03-04 19:18:56.000000000 -0500
***************
*** 3,9 ****
  
  use Legacy::Graphics::Browser::Util 'shellwords';
  use CGI::Session;
! use constant COOKIE_NAME => 'gbrowse_sess';
  
  sub new {
    my $class    = shift;
--- 3,9 ----
  
  use Legacy::Graphics::Browser::Util 'shellwords';
  use CGI::Session;
! use constant COOKIE_NAME => 'gbrowse2_sess';
  
  sub new {
    my $class    = shift;
