--- gbrowse_img.orig	2012-03-06 16:33:44.000000000 -0500
+++ gbrowse_img	2012-03-06 16:40:52.000000000 -0500
@@ -203,6 +203,12 @@
     my $format   = param('format') || 'GD';
     my $flip     = param('flip')   || 0;
     my $embed    = param('embed');
+    # eupathdb patches for image map name 3/6/12
+    my $hmap     = param('hmap');
+
+    if (my @regions = param('h_region')) {
+       $render->state->{h_region} = \@regions;
+    } 
 
     if (my @regions = param('h_region')) {
 	$render->state->{h_region} = \@regions;
@@ -220,7 +226,7 @@
 
     my ($img_data,$map) = $render->region->feature_count > 1 
                                  ? $self->render_multiple($renderer,$format,$flip,$embed)
-	                         : $self->render_tracks  ($renderer,$format,$flip,$embed);
+	                         : $self->render_tracks  ($renderer,$format,$flip,$embed,$hmap);
 
  
     my $seg   = $render->segment;
@@ -230,14 +236,14 @@
         my $url    = $renderer->source->generate_image($img_data);
 	my $js = $render->data_source->globals->js_url;
 	my @scripts = map { {src=>"$js/$_"} }
-	                  qw(balloon.js balloon.config.js yahoo-dom-event.js);
+	                  qw(prototype.js GBox.js balloon.js balloon.config.js yahoo-dom-event.js);
 	print $self->header(-type=>'text/html');
 	print start_html(-script=>\@scripts),
 	      $render->render_balloon_settings,
 	      img(
 		  {
 		      -src    => $url,
-		      -usemap => "#gbrowse2_img_map"}
+          -usemap => "#". ($hmap || "gbrowse2_img") . "_map"}
 		  ),
 	      $map,
 	      end_html();
@@ -294,7 +300,7 @@
 
 sub render_tracks {
     my $self = shift;
-    my ($renderer,$format,$flip,$embed) = @_;
+    my ($renderer,$format,$flip,$embed,$hmap) = @_;
 
     my $render   = $self->render;
 
@@ -333,7 +339,7 @@
     my @map_data        = map {$_->{map}} @image_data;
 
     my $img_data  = $self->consolidate_images(\@gds);
-    my $map       = $self->consolidate_maps  (\@map_data, \@gds) if $embed;
+    my $map       = $self->consolidate_maps  (\@map_data, \@gds, $hmap) if $embed;
 
     return ($img_data,$map);
 
@@ -530,10 +536,11 @@
 
 sub consolidate_maps {
     my $self = shift;
-    my ($maps,$gds) = @_;
+    my ($maps,$gds,$hmap) = @_;
 
     my $offset = 0;
-    my @integrated_list = 'gbrowse2_img';
+    my @integrated_list = $hmap || 'gbrowse2_img';
+
     for (my $i=0;$i<@$maps;$i++) {
 	my $data = $maps->[$i];
 	shift @$data;
