--- MANIFEST.dist	2012-02-06 09:46:58.000000000 -0500
+++ MANIFEST	2012-02-06 09:48:33.000000000 -0500
@@ -35,20 +35,8 @@
 conf/detail_select_menu.conf
 conf/enzymes.txt
 conf/GBrowse.conf
-conf/languages/de.pm
 conf/languages/en.pm
-conf/languages/es.pm
-conf/languages/fr.pm
-conf/languages/is.pm
-conf/languages/it.pm
-conf/languages/ja.pm
-conf/languages/ko.pm
-conf/languages/nl.pm
-conf/languages/pl.pm
 conf/languages/POSIX.pm
-conf/languages/ru.pm
-conf/languages/zh.pm
-conf/languages/zh_cn.pm
 conf/log4perl.conf
 conf/MobyServices/b64_encoded_gif_renderer.pm
 conf/MobyServices/b64_encoded_jpeg_renderer.pm
