--- PrimerDesigner.pm.orig	2012-12-10 15:15:25.000000000 -0500
+++ PrimerDesigner.pm	2012-12-10 15:21:53.000000000 -0500
@@ -86,13 +86,13 @@
 use Carp qw/cluck/;
 
 use constant BINARY            => 'primer3';
-use constant BINPATH           => '/usr/local/bin';
+use constant BINPATH           => "$ENV{GUS_HOME}/bin";
 use constant METHOD            => 'local';
 use constant IMAGE_PAD         => 25;
 use constant MAXRANGE          => 300;
 use constant IMAGEWIDTH        => 800;
 use constant DEFAULT_SEG_SIZE  => 10000;
-use constant STYLE             => '/gbrowse2/css/gbrowse.css';
+use constant STYLE             => '/assets/css/eupath_gbrowse.css';
 
 #use vars '@ISA';
 use vars qw/@ISA $PNG $CALL/;
@@ -1265,6 +1265,9 @@
 
   my $sr = $conf->{size_range} || '';
 
+	my $primer_min_tm = "57.0";
+	$primer_min_tm = "50.0" if ($ENV{PROJECT_ID} =~ /PlasmoDB/i);
+
   my %table = (
     b(qq(<a name="PRIMER_NUM_RETURN_INPUT" target="_new" href="$help\#PRIMER_NUM_RETURN">
        Primer sets:</a>)
@@ -1279,7 +1282,7 @@
     b(qq(<a name="PRIMER_OPT_TM_INPUT" target="_new" href="$help\#PRIMER_TM">
           Primer Tm</a>)
     ),
-    qq(Min. <input type="text" size="4" name="PRIMER_MIN_TM" value="57.0">
+    qq(Min. <input type="text" size="4" name="PRIMER_MIN_TM" value="$primer_min_tm">
        Opt. <input type="text" size="4" name="PRIMER_OPT_TM" value="60.0">
        Max. <input type="text" size="4" name="PRIMER_MAX_TM" value="63.0">),
     b(qq(<a name="PRIMER_PRODUCT_SIZE_RANGE" href="javascript:void(0)"
