--- RegionSearch.pm.orig	2012-03-07 12:28:13.000000000 -0500
+++ RegionSearch.pm	2012-03-07 12:31:03.000000000 -0500
@@ -721,7 +721,7 @@
 sub features {
     my $self    = shift;
     my $segment = $self->segment;
-    $self->db->features(-seq_id => $segment->seq_id,
+    $segment->features(-seq_id => $segment->seq_id,
 			-start  => $segment->start,
 			-end    => $segment->end,
 #			-class  => eval {$segment->class} || 'Sequence',
