--- SequenceDumper.pm.orig	2012-03-05 19:33:18.000000000 -0500
+++ SequenceDumper.pm	2012-03-05 19:35:07.000000000 -0500
@@ -215,8 +215,13 @@
   print "##See http://www.sanger.ac.uk/Software/formats/GFF/\n";
   print "##NOTE: Selected features dumped.\n";
   my @feature_types = $self->selected_features;
+
+  # eupathdb patch 3/5/2012
+  my @sqlName = $self->selected_sqlName;
+  my @sqlParam = $self->selected_sqlParam; 
+
   $segment->absolute(0);
-  my $iterator = $segment->get_seq_stream(-types => \@feature_types) or return;
+  my $iterator = $segment->get_seq_stream(-types => \@feature_types, -sqlName=>\@sqlName, -sqlParam => \@sqlParam) or return;
   while (my $f = $iterator->next_seq) {
       eval{$f->version(3)};
       my $s =$f->gff_string(1);
@@ -224,6 +229,12 @@
       print $s,"\n";
   }
 
+  print "##FASTA\n";
+  my $seq = $segment->seq;
+  $seq =~ s/(.{1,60})/$1\n/g;
+  print ">".$segment->ref,":",$segment->start,"..",$segment->stop, "\n";
+  print $seq;
+
   print end_pre() if $html;
   print end_html() if $html;
 }
