diff -rupN cgi-bin.dist/gbrowse cgi-bin/gbrowse
--- cgi-bin.dist/gbrowse	2012-03-04 15:37:22.000000000 -0500
+++ cgi-bin/gbrowse	2012-03-04 15:37:57.000000000 -0500
@@ -8,7 +8,7 @@ use CGI;
 use Bio::Graphics::Browser2;
 use Bio::Graphics::Browser2::Render::HTML;
 
-umask 022;
+umask 0;
 
 my $fcgi    = Bio::Graphics::Browser2::Render->fcgi_request;
 my $init;
diff -rupN cgi-bin.dist/gbrowse_img cgi-bin/gbrowse_img
--- cgi-bin.dist/gbrowse_img	2012-03-04 15:37:22.000000000 -0500
+++ cgi-bin/gbrowse_img	2012-03-04 15:38:59.000000000 -0500
@@ -23,7 +23,7 @@ use Digest::MD5 'md5_hex';
 # h_region  region(s) to hilight
 # h_feat    feature(s) to hilight
 
-umask 022;
+umask 0;
 
 my $fcgi = Bio::Graphics::Browser2::Render->fcgi_request;
 my $modperl = $ENV{MOD_PERL};
