--- wiggle_xyplot.pm.orig	2012-12-12 15:21:12.000000000 -0500
+++ wiggle_xyplot.pm	2012-12-12 15:27:15.000000000 -0500
@@ -212,7 +212,11 @@
 	    $x2        = $right - ($x2-$left) if $flip;
  
 	    my $color = $score > $midpoint ? $positive : $negative;
+			if($score < 0) { 
+	    [int($x1+0.5),int($y2+0.5),int($x2+0.5),int($y1+0.5),$color,$lw];
+			} else {
 	    [int($x1+0.5),int($y1+0.5),int($x2+0.5),int($y2+0.5),$color,$lw];
+			}
 	} else {
 	    ();
 	}
