--- gbrowse	2006-06-20 21:50:44.000000000 -0400
+++ gbrowse.presets	2006-06-20 21:49:09.000000000 -0400
@@ -31,6 +31,7 @@
 use CGI qw(:standard unescape escape escapeHTML center *table *dl *TR *td);
 use CGI::Toggle;
 use CGI::Cookie;
+use CGI::Session;
 use vars qw($CONFIG $MAX_SEGMENT $DEFAULT_SEGMENT
 	    $HEADER $HTML $UA $VERSION $BIOGRAPHICS_VERSION $CONF_DIR
 	    $PLUGINS $UPLOADED_SOURCES $REMOTE_SOURCES
@@ -254,7 +255,12 @@
 my $cookies;
 if (param('reset')) {
   $cookies = reset_cookies($page_settings);
-} else {
+} elsif (param('save')) {
+  $cookies = save_user_settings($page_settings,$session, param('setting_name'), param('setting_options'));
+} elsif (param('delete')) {
+	delete_settings(param('delete'), $session);
+}
+else {
   $cookies = [$cookie];
 }
 
@@ -413,6 +419,8 @@
 		TR(td({-align=>'left'},
 		      a({-href=>"?name=$page_settings->{name};h_feat=_clear_;h_region=_clear_"},
 			font({-size=>-2},$CONFIG->tr('Clear_highlighting')))),
+			td({-align=>'left', -valign=>'middle'},
+					font({-size=>-2},$CONFIG->tr('CONSERVATION'))),
 		   td({-align=>'right'},
 		      b(submit(-name => $CONFIG->tr('Update'))))));
 
@@ -432,6 +440,7 @@
   print html_frag('html6',$segment,$page_settings);
   print end_form();
 
+	print preset_table($page_settings, $session);
   unless ($CONFIG->section_setting('upload_tracks') eq 'off') {
     print start_multipart_form(-name=>'externalform');
     print external_table($page_settings,$feature_files);
@@ -1135,6 +1144,158 @@
   return $category ||= $CONFIG->tr('GENERAL');
 }
 
+sub preset_table {
+
+	my $settings = shift;
+	my $session = shift;
+
+	my @labels = @{$settings->{tracks}};
+	my %preset_tracks;
+
+	foreach (@labels) {
+		my $preset = $CONFIG->setting($_ => 'preset');
+		if($preset) { 
+			my @sets = split /\s+/, $preset;
+			foreach my $name (@sets) {
+				push @{$preset_tracks{$name}}, $_;
+			} 
+		} 
+	}
+
+	my @preset_lists = ();
+	while(my ($k,$v) = each %preset_tracks) {
+	
+		my $str = ""; 
+		foreach(@$v) { 
+			$str .= $_. "-"; 
+		} 
+		$str =~ s/\-$//ig; 
+		$str = qq|<a href="?name=$page_settings->{name};label=$str">$k</a><br />|; 
+		push @preset_lists, $str; 
+	}
+
+	my $hash = $session->session->param('user_tracks');
+	my $track_lists .= qq(<table border=0 >);
+	my $region_lists .= qq(<table border=0 >);
+	foreach(keys %$hash) {
+
+    my $my_setting = $$hash{$_};
+		my ($tracks, $name, $region);
+    if($my_setting =~ /^(.*?)######(.*?)$/) {
+			$tracks = $1;
+			$name = $2;
+			$track_lists .= qq|<tr><td><a href="?delete=$_"><font color=red>Delete</font></a>&nbsp;&nbsp;<a href="?name=$page_settings->{name};label=$tracks">$name</a></td></tr> |; 
+		} else {
+			$region_lists .= qq|<tr><td><a href="?delete=$_"><font color=red>Delete</font></a>&nbsp;&nbsp;<a href="?name=$my_setting">$my_setting</a></td></td></tr> |; 
+		}
+	}
+	$track_lists .= qq(</table>);
+	$region_lists .= qq(</table>);
+
+	autoEscape(0);
+
+  my @options = split /\s+/,$CONFIG->setting('setting options');
+  @options = ('Tracks','Region') unless @options;
+
+  my $save = join('',
+	        start_form(-name=>'userform', -method=>'GET'),
+					table({-class=>'searchbody',-border=>0,-width=>'100%'}, 
+					TR( td({-align=>'right',-valign=>'bottom'},
+						 submit(-value => $CONFIG->tr('SAVE_SETTINGS'), -name=>'save'),
+						radio_group( -name=>'setting_options',
+										 -values=>\@options,
+										 -default=>$settings->{option},
+										 -override=>1,
+									 ),
+					 textfield(-name  => 'setting_name',
+			       -value => 'Track/Region Description',
+			       -size  => 25,
+			       -override=>1,
+			      )))),
+					end_form());
+  
+	my $content = 
+		table({-class=>'searchbody',-border=>0,-width=>'100%'}, 
+		TR(
+			th({-align=>'left'}, 'Preset Track Configurations'),
+		),
+		TR(
+			td(@preset_lists,),
+			),
+		TR( 
+			th({-align=>'left'}, 'My Presets/Saved Tracks'),
+		),
+		TR(
+			td($track_lists),
+		),
+		TR(
+			th({-align=>'left'}, 'My Regions of Interest'),
+		),
+		TR(
+			td($region_lists),
+		) 
+		);
+
+ 	return toggle('PRESET_TRACKS',$content.$save);
+}
+
+# Save current user settings (tracks/regions)
+sub save_user_settings {
+
+  my $settings = shift; 
+	my $session = shift;
+	my $name = shift;
+	my $type = shift;
+
+	my $region = $settings->{name};
+	$name  = $name || 'My Setting'; 
+	my @labels = $CONFIG->labels; 
+	my $str = ""; 
+	foreach (@labels) { 
+		if( $settings->{features}{$_}{visible} == 1) { 
+			$str .= $_."-"; 
+		} 
+	} 
+	$str =~ s/\-$//i; 
+
+  if($type eq 'Tracks') {
+		$str .= "######". $name;
+	} elsif ($type eq 'Region') {
+		$str =  "$region ($name)";
+	}
+	
+	my $user_session = new CGI::Session("driver:File", undef, {Directory=>'/tmp'});
+
+	my $hash = $session->session->param('user_tracks');
+	my $uid = $user_session->id;
+	if ($hash) {
+	  my $flag = 0;
+	  foreach(keys %$hash) {
+		  if($str eq $$hash{$_}) {
+				$flag = 1;
+			}
+		}
+		if($flag == 0) {
+		$$hash{$uid} = $str;
+		}
+		$session->session->param(-name => 'user_tracks', -value => $hash);
+	} else {
+		my %hash = ($uid => $str);
+		$session->session->param(-name => 'user_tracks', -value => \%hash);
+	}
+}
+
+sub delete_settings {
+
+  my $uid = shift;
+	my $session = shift;
+
+	my $hash = $session->session->param('user_tracks');
+	if (exists $$hash{$uid}) {
+		delete($$hash{$uid}); 
+		$session->session->param(-name => 'user_tracks', -value => $hash);
+	} 
+}
 
 sub settings_table {
   my $settings = shift;
@@ -1240,7 +1401,8 @@
   my $upload_table = upload_table($settings,$feature_files);
   my $das_table    = das_table($settings,$feature_files);
   toggle('Upload_tracks',
-	 table({-width=>'100%',-class=>'searchbody'},
+	 #table({-width=>'100%',-class=>'searchbody',-border=>1},
+	 table({-width=>'100%',-class=>'searchtitle',-border=>1},
 	       TR(td($upload_table,$das_table))));
 }
 
