--- gbrowse_img	2005-12-06 16:03:55.000000000 -0500
+++ cgi-bin/gbrowse_img	2006-02-01 10:59:44.000000000 -0500
@@ -185,20 +185,24 @@
 if ($embed) {
   my $url             = $CONFIG->generate_image($img);
   my ($width,$height) = $img->getBounds;
-  my $img             = img({-src=>$url,-align=>'middle',-usemap=>'#hmap',-width=>$width,
+  my $hmap            = param("hmap") || "hmap";
+  my $img             = img({-src=>$url,-align=>'middle',-usemap=>"#$hmap",-width=>$width,
 			     -height=>$height,-border=>0,-name=>'detailedView',-alt=>'detailed view'});
   my $source          = $CONFIG->source;
   my $map;
+  my $base            = param("base") || "../../gbrowse";
 
   if ($detailed_view) {
     my $ruler           = shift @$boxes;
     $map                = $CONFIG->make_map($boxes,0,$panel,$tracks);
     my $coords          = join ',',@{$ruler}[1,2,3,4];
-    my $gbrowse         = "../../gbrowse/$source/?name=$segments[0]"; # dgg; was  "../gbrowse/$source?"
+    my $gbrowse         = "${base}/$source/?name=$segments[0]"; # dgg; was  "../gbrowse/$source?"
     my $rect            = qq(<area shape="rect" coords="$coords" href="$gbrowse" title="Browse Region" />);
     $map                =~ s!</map>!$rect</map>!;
+    $map                =~ s!name="hmap"!name="$hmap"!;
+    $map                =~ s!id="hmap"!id="$hmap"!;
   } else {
-    $map                = my_make_map($boxes,$source,scalar @add_features);
+    $map                = my_make_map($boxes,$source,scalar(@add_features),$hmap,$base);
   }
 
   $map =~ s/href/target="_top" href/g;  # add appropriate target tags
@@ -447,9 +451,9 @@
 }
 
 sub my_make_map {
-  my ($boxes,$source,$add_features) = @_;
-  my $url = "../../gbrowse/$source/";  # dgg; was"../gbrowse/$source";
-  my $html = qq(<map name="hmap" alt="imagemap" />\n);
+  my ($boxes,$source,$add_features,$hmap,$base) = @_;
+  my $url = "$base/$source/";  # dgg; was"../gbrowse/$source";
+  my $html = qq(<map name="$hmap" alt="imagemap" />\n);
   foreach (@$boxes){
     my $feature = $_->[0];
     my $ref     = $_->[6];  # consolidate_images() sets this slot
