--- Database.pm.orig	2011-09-30 17:28:56.000000000 -0400
+++ Database.pm	2011-09-30 17:30:05.000000000 -0400
@@ -220,6 +220,8 @@
                       . " OR title LIKE "           . $uploadsdb->quote("%".$searchterm."%") . ")" if $searchterm;
     
     my $rows = $uploadsdb->selectcol_arrayref($sql);
+		# a tempory patch to work around userdb EuPathDB 9/28/2011
+		$rows = [];
     return @$rows;
 }
 
