--- GBrowse.conf.orig	2011-10-27 14:00:03.000000000 -0400
+++ GBrowse.conf	2011-10-27 14:00:07.000000000 -0400
@@ -5,7 +5,7 @@
 [GENERAL]
 config_base            = /var/www/ToxoDB/toxo.mheiges/conf/gbrowse.conf   # overridden by environment variable GBROWSE_CONF
 htdocs_base            = /var/www/ToxoDB/toxo.mheiges/html/gbrowse
-url_base               = /gbrowse2
+url_base               = /gbrowse
 tmp_base               = /var/www/ToxoDB/toxo.mheiges/html/gbrowse/tmp
 persistent_base        = /var/www/ToxoDB/toxo.mheiges/cgi-lib/persistent
 userdata_base          = /var/www/ToxoDB/toxo.mheiges/cgi-lib/persistent/userdata
@@ -234,4 +234,19 @@
 description   = Population Display Demo
 path          = pop_demo.conf
 
+[plasmodb]
+description   = PlasmoDB GBrowse v2.39
+path          = plasmodb.conf
+
+[plasmodbaa]
+description   = PlasmoDB PBrowse v2.39
+path          = plasmodbaa.conf
+
+[toxodb]
+description   = ToxoDB GBrowse v2.39
+path          = toxodb.conf
+
+[toxodbaa]
+description   = ToxoDB PBrowse v2.39
+path          = toxodbaa.conf
 
