--- GBrowse.conf.orig	2011-12-13 11:27:04.000000000 -0500
+++ GBrowse.conf	2011-12-13 11:56:30.000000000 -0500
@@ -97,7 +97,7 @@
 
 balloon tips        = 1
 titles are balloons = 1
-plugins             = FastaDumper RestrictionAnnotator SequenceDumper TrackDumper
+plugins             = SequenceDumper RestrictionAnnotator FastaDumper Submitter
 overview grid       = 0
 region grid         = 0
 detail grid         = 1
@@ -160,8 +160,8 @@
 # "user_accounts" is true, then GBrowse
 # will attempt to use its internal user accounts database
 # to authenticate and/or register users.
-user_accounts               = 1
-user_accounts_registration  = 1
+user_accounts               = 0
+user_accounts_registration  = 0
 user_accounts_openid        = 0
 
 # Path to the database -- you will need to create this database and grant all
@@ -234,4 +234,4 @@
 description   = Population Display Demo
 path          = pop_demo.conf
 
-
+#include "eupath_datasource.conf"
