--- RegionSearch.pm.orig	2011-12-16 12:10:07.000000000 -0500
+++ RegionSearch.pm	2011-12-16 14:11:35.000000000 -0500
@@ -396,6 +396,9 @@
     warn "dbs = @dbids" if DEBUG;
     my %seenit;
 
+    # prevent uninit variable warnings here
+    $args->{-name} ||= '';
+
     for my $dbid (@dbids) {
 	warn "searching in ",$dbid if DEBUG;
 	my $db = $self->source->open_database($dbid);
@@ -719,7 +722,9 @@
 sub features {
     my $self    = shift;
     my $segment = $self->segment;
-    $self->db->features(-seq_id => $segment->seq_id,
+    # eupathdb patch 12/5/2011 sqlparam and sqlname
+    #$self->db->features(-seq_id => $segment->seq_id,
+    $segment->features(-seq_id => $segment->seq_id,
 			-start  => $segment->start,
 			-end    => $segment->end,
 			-class  => eval {$segment->class} || 'Sequence',
