--- RenderPanels.pm.orig	2011-09-28 14:39:38.000000000 -0400
+++ RenderPanels.pm	2011-09-29 14:20:05.000000000 -0400
@@ -1465,6 +1465,15 @@
 
 		    warn "track_setup($label): ",time()-$time," seconds " if BENCHMARK;
 
+        # EuPathDB patches 9/28/2011
+        my @array = @$track_args;
+        my ($sqlParam, $sqlName);
+        my ($index) = grep { $array[$_] =~ /sqlParam/i } 0..$#array;
+        $sqlParam = $array[$index+1] if $index;
+        my ($index2) = grep { $array[$_] =~ /sqlName/i } 0..$#array;
+        $sqlName = $array[$index2+1] if $index;
+				# end EuPathDB patches 
+
 		    # == populate the tracks with feature data ==
 		    $self->add_features_to_track(
 			-labels    => [ $label, ],
@@ -1472,6 +1481,8 @@
 			-filters   => $filters,
 			-segment   => $segment,
 			-fsettings => $settings->{features},
+      -sqlParam  => [$sqlParam],
+      -sqlName   => [$sqlName], 
 			);
 
 		    warn "add_features($label): ",time()-$time," seconds " if BENCHMARK;
@@ -1595,6 +1606,8 @@
   my $tracks          = $args{-tracks}    or die "programming error";
   my $filters         = $args{-filters}   or die "programming error";
   my $fsettings       = $args{-fsettings} or die "programming error";
+  my $sqlParam        = $args{-sqlParam} or die "programming error";
+  my $sqlName         = $args{-sqlName} or die "programming error";
 
   warn "[$$] add_features_to_track @{$args{-labels}}" if DEBUG;
 
@@ -1636,7 +1649,7 @@
 	  $iterator2dbid{$iterator} = $source->db2id($db);
       }
 
-      if (@full_types && (my $iterator = $self->get_iterator($db2db{$db},$segment,\@full_types))) {
+      if (@full_types && (my $iterator = $self->get_iterator($db2db{$db},$segment,\@full_types,$sqlName,$sqlParam))) {
 	  $iterators{$iterator}     = $iterator;
 	  $iterator2dbid{$iterator} = $source->db2id($db);
       }
@@ -1849,7 +1862,7 @@
 
 sub get_iterator {
   my $self = shift;
-  my ($db,$segment,$feature_types,$max) = @_;
+  my ($db,$segment,$feature_types,$sqlName,$sqlParam,$max) = @_;
 
   # The Bio::DB::SeqFeature::Store database supports correct
   # semantics for directly retrieving features that overlap
@@ -1863,6 +1876,8 @@
 		  -start  => $segment->start,
 		  -end    => $segment->end,
 		  -max_features => $max,  # some adaptors allow this
+      -sqlParam => $sqlParam,
+      -sqlName  => $sqlName,
 	  );
       return $db->get_seq_stream(@args);
   }
@@ -1880,7 +1895,7 @@
     return;
   }
 
-  return $db_segment->get_seq_stream(-type=>$feature_types);
+  return $db_segment->get_seq_stream(-type=>$feature_types,-sqlName=>$sqlName,-sqlParam=>$sqlParam);
 }
 
 sub get_summary_iterator {
@@ -1999,7 +2014,9 @@
 	      -pad_top      => $image_class->gdMediumBoldFont->height+2,
               -pad_bottom   => 3,
 	      -image_class  => $image_class,
-	      -postgrid     => $postgrid,
+	      # EuPathDB patch 9/28/2011
+	      #-postgrid     => $postgrid,
+	      -postgrid     => $section eq 'detail' ? $source->global_setting('postgrid') : '',
 	      -background   => $args->{background} || '',
 	      -truecolor    => $source->global_setting('truecolor') || 0,
 	      -map_fonts_to_truetype    => $source->global_setting('truetype') || 0,
