--- SequenceDumper.pm.orig	2012-02-02 12:31:03.000000000 -0500
+++ SequenceDumper.pm	2012-02-02 15:37:43.000000000 -0500
@@ -95,7 +95,7 @@
 
 @ISA = qw(Bio::Graphics::Browser2::Plugin);
 
-sub name { "Sequence File" }
+sub name { "Track Data" }
 sub description {
   p("The Sequence dumper plugin dumps out the currently displayed genomic segment",
     "in the requested format.").
@@ -275,8 +275,13 @@
   print "##See http://www.sanger.ac.uk/Software/formats/GFF/\n";
   print "##NOTE: Selected features dumped.\n";
   my @feature_types = $self->selected_features;
+
+  # eupathdb patch 12/5/2011
+  my @sqlName = $self->selected_sqlName;
+  my @sqlParam = $self->selected_sqlParam; 
+
   $segment->absolute(0);
-  my $iterator = $segment->get_seq_stream(-types => \@feature_types) or return;
+  my $iterator = $segment->get_seq_stream(-types => \@feature_types, -sqlName=>\@sqlName, -sqlParam => \@sqlParam) or return;
   while (my $f = $iterator->next_seq) {
       eval{$f->version(3)};
       my $s =$f->gff_string(1);
@@ -284,6 +289,12 @@
       print $s,"\n";
   }
 
+  print "##FASTA\n";
+  my $seq = $segment->seq;
+  $seq =~ s/(.{1,60})/$1\n/g;
+  print ">".$segment->ref,":",$segment->start,"..",$segment->stop, "\n";
+  print $seq;
+
   print end_pre() if $html;
   print end_html() if $html;
 }
