--- SequenceDumper.pm.orig	2011-12-16 15:37:21.000000000 -0500
+++ SequenceDumper.pm	2011-12-16 16:35:48.000000000 -0500
@@ -95,7 +95,7 @@
 
 @ISA = qw(Bio::Graphics::Browser2::Plugin);
 
-sub name { "Sequence File" }
+sub name { "Track Data" }
 sub description {
   p("The Sequence dumper plugin dumps out the currently displayed genomic segment",
     "in the requested format.").
@@ -275,8 +275,13 @@
   print "##See http://www.sanger.ac.uk/Software/formats/GFF/\n";
   print "##NOTE: Selected features dumped.\n";
   my @feature_types = $self->selected_features;
+
+  # eupathdb patch 12/5/2011
+  my @sqlName = $self->selected_sqlName;
+  my @sqlParam = $self->selected_sqlParam; 
+
   $segment->absolute(0);
-  my $iterator = $segment->get_seq_stream(-types => \@feature_types) or return;
+  my $iterator = $segment->get_seq_stream(-types => \@feature_types, -sqlName=>\@sqlName, -sqlParam => \@sqlParam) or return;
   while (my $f = $iterator->next_seq) {
       eval{$f->version(3)};
       my $s =$f->gff_string(1);
